### PR TITLE
Fix MultiInfo parser + serialization bug

### PR DESCRIPTION
**Pull request by jared-barocsi**
This PR addresses the issue highlighted in #2253: in particular, the firrtl parser can now reconstruct a `MultiInfo` structure that was previously serialized and compressed by the algorithm in #2212. This only results in a single level `MultiInfo` due to the flattening step of serialization, so it can not recreate a `MultiInfo` of `MultiInfo`s or any such recursive source locator object.

### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [ ] Did you add at least one test demonstrating the PR? **(Test is added in another PR that needs to be merged.)**
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

bug fix

#### API Impact

No change

#### Backend Code Generation Impact

Should have no impact.

#### Desired Merge Strategy

Squash and merge

#### Release Notes
Firrtl parser can now handle parsing a serialized, compressed source locator and return a `MultiInfo` object consisting of `FileInfo`s.

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (1.2.x, 1.3.0, 1.4.0) ?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you mark as `Please Merge`?

### CODE REVIEW

**Comment by ekiwi** [comment]
Looking good. Could you please add the test from https://github.com/chipsalliance/firrtl/pull/2253 to show that your PR solves the issue?

File: src/test/scala/firrtlTests/InfoSpec.scala [hunk 1]
```diff
@@ -321,4 +321,17 @@ class InfoSpec extends FirrtlFlatSpec with FirrtlMatchers {
     for (line <- check)
       result should containLine(line)
   }
+
+  "Source Locators" should "not loose information when going through serialization + parsing" in {
```
Comment (azidar) [inline src/test/scala/firrtlTests/InfoSpec.scala:325]:
```suggestion
  "Source Locators" should "not lose information when going through serialization + parsing" in {
```

File: src/main/scala/firrtl/Visitor.scala [hunk 2]
```diff
@@ -58,6 +58,56 @@ class Visitor(infoMode: InfoMode) extends AbstractParseTreeVisitor[FirrtlNode] w
...
+
+      // Regular expression to match against individual column numbers in each line:col or line:{col1,col2}
+      // descriptor.
+      val splitColDescriptors = """\d+""".r
+
+      val matches = splitCompressedInfo.findAllIn(escaped)
+
+      // Grab each File.format line:col token from the input string
+      splitCompressedInfo.findAllIn(escaped).matchData.foreach { info =>
+        {
```
Comment (azidar) [inline src/main/scala/firrtl/Visitor.scala:82]: Unnecessary curly braces here

File: src/main/scala/firrtl/Visitor.scala [hunk 3]
```diff
@@ -58,6 +58,56 @@ class Visitor(infoMode: InfoMode) extends AbstractParseTreeVisitor[FirrtlNode] w
...
+          Option(info.group(1)) match {
+            // If there were no subgroups, the regex matched against a non-conforming source locator
+            // pattern, so do not process it
+            case None => out = out :+ ir.FileInfo.fromEscaped(info.toString)
+            case Some(file) => {
```
Comment (azidar) [inline src/main/scala/firrtl/Visitor.scala:87]:
```suggestion
            case Some(file) =>
```
Again, unnecessary curly braces.

File: src/main/scala/firrtl/Visitor.scala [hunk 4]
```diff
@@ -58,6 +58,56 @@ class Visitor(infoMode: InfoMode) extends AbstractParseTreeVisitor[FirrtlNode] w
...
+              val lineDescriptors = info.group(2)
+              // Grab each line:col values from the separated (compressed) FileInfo.
+              splitLineDescriptors.findAllIn(lineDescriptors).matchData.foreach { lineDescriptor =>
+                {
```
Comment (azidar) [inline src/main/scala/firrtl/Visitor.scala:91]: Again :)

**Review by azidar — APPROVED** [review]
LGTM, just minor syntax nits.

File: src/test/scala/firrtlTests/InfoSpec.scala (hunk 1 above)
Comment (ekiwi) [reply src/test/scala/firrtlTests/InfoSpec.scala:325]: my bad :(

**Pull request merged**